File: src/main/java/org/oewntk/model/MapFactory.java
/*
 * Copyright (c) 2021. Bernard Bou.
 */

package org.oewntk.model;

import java.util.Collection;
import java.util.Map;
import java.util.TreeMap;
import java.util.function.BinaryOperator;
import java.util.function.Function;
import java.util.function.Supplier;

import static java.util.stream.Collectors.toMap;

/**
 * Map factory
 */
public class MapFactory
{
	private static final boolean LOG_DUPLICATE_VALUES = false;

	// G E N E R I C   M A P   F A C T O R Y

	/**
	 * Supply 'keep' merging function
	 *
	 * @param <V> type of element
	 */
	private static class KeepMergingSupplier<V> implements Supplier<BinaryOperator<V>>
	{
		@Override
		public BinaryOperator<V> get()
		{
			return (existing, replacement) -> {
				if (existing.equals(replacement))
				{
					if (LOG_DUPLICATE_VALUES)
					{
						Tracing.psErr.printf("[W] Duplicate values %s and %s, keeping first%n", existing, replacement);
					}
					//throw new IllegalArgumentException(existing + "," + replacement);
				}
				return existing;
			};
		}
	}

	/**
	 * Supply 'replace' merging function
	 *
	 * @param <V> type of element
	 */
	private static class ReplaceMergingSupplier<V> implements Supplier<BinaryOperator<V>>
	{
		@Override
		public BinaryOperator<V> get()
		{
			return (existing, replacement) -> {
				if (existing.equals(replacement))
				{
					if (LOG_DUPLICATE_VALUES)
					{
						Tracing.psErr.printf("[W] Duplicate values %s and %s, replacing first%n", existing, replacement);
					}
					//throw new IllegalArgumentException(existing + "," + replacement);
				}
				return replacement;
			};
		}
	}

	/**
	 * Make map
	 *
	 * @param things           elements
	 * @param groupingFunction map element to key
	 * @param <K>              type of key
	 * @param <V>              type of element
	 * @return elements mapped by key
	 */
	public static <K, V> Map<K, V> map(final Collection<V> things, final Function<V, K> groupingFunction)
	{
		var mergingFunction = new KeepMergingSupplier<V>().get();
		return map(things, groupingFunction, mergingFunction);
	}

	/**
	 * Make map
	 *
	 * @param things           elements
	 * @param groupingFunction map element to key
	 * @param mergingFunction  merging function for existing and replacement elements
	 * @param <K>              type of key
	 * @param <V>              type of element
	 * @return elements mapped by key
	 */
	public static <K, V> Map<K, V> map(final Collection<V> things, final Function<V, K> groupingFunction, final BinaryOperator<V> mergingFunction)
	{
		return things.stream() //
				.collect(toMap(groupingFunction, //
						Function.identity(), //
						mergingFunction, //
						TreeMap::new));
	}

	// G E N E R I C   M A P   F A C T O R Y

	/**
	 * Senses by id
	 *
	 * @param senses senses
	 * @return senses mapped by id
	 */
	public static Map<String, Sense> sensesById(final Collection<Sense> senses)
	{
		BinaryOperator<Sense> mergingFunction = (existing, replacement) -> {

			Sense merged = replacement.getLex().isCased() ? (existing.getLex().isCased() ? existing : replacement) : existing;
			if (existing.equals(replacement))
			{
				if (LOG_DUPLICATE_VALUES)
				{
					Tracing.psErr.printf("[W] Duplicate values %s and %s, merging to %s%n", existing, replacement, merged);
				}
			}
			return merged;
		};
		return map(senses, Sense::getSensekey, mergingFunction);
	}

	/**
	 * Synsets by id
	 *
	 * @param synsets synsets
	 * @return synsets mapped by id
	 */
	public static Map<String, Synset> synsetsById(final Collection<Synset> synsets)
	{
		return map(synsets, Synset::getSynsetId);
	}
}

File: src/main/java/org/oewntk/model/MapFactory.kt
/*
 * Copyright (c) 2021-2024. Bernard Bou.
 */
package org.oewntk.model

import java.io.Serializable

/**
 * Map factory
 */
object MapFactory {

    private const val LOG_DUPLICATE_VALUES = false

    /**
     * A generic function that returns a Serializable Comparator for any Comparable type T
     */
    fun <T : Comparable<T>> serializableNaturalOrder(): Comparator<T> =
        object : Comparator<T>, Serializable {
            override fun compare(o1: T, o2: T): Int = o1.compareTo(o2)
        }

    /**
     * Make map
     *
     * @param things           elements
     * @param groupingFunction map element to key
     * @param K                type of key
     * @param V                type of element
     * @return elements mapped by key
     */
    fun <K : Comparable<K>, V> map(
        things: Sequence<V>,
        groupingFunction: (V) -> K,
    ): Map<K, V> {
        return map(things, groupingFunction, keepMerging())
    }

    /**
     * Make map (alt)
     *
     * @param things           elements
     * @param groupingFunction map element to key
     * @param K                type of key
     * @param V                type of element
     * @return elements mapped by key
     */
    fun <K : Comparable<K>, V> mapAlt(
        things: Sequence<V>,
        groupingFunction: (V) -> K,
    ): Map<K, V> {
        return things
            .associateBy { groupingFunction(it) }
            .toSortedMap(serializableNaturalOrder())
    }

    /**
     * Make map
     *
     * @param things           elements
     * @param groupingFunction map element to key
     * @param mergingFunction  merging function for existing and replacement elements
     * @param K                type of key
     * @param V                type of element
     * @return elements mapped by key
     */
    fun <K : Comparable<K>, V> map(
        things: Sequence<V>,
        groupingFunction: (V) -> K,
        mergingFunction: (V, V) -> V,
    ): Map<K, V> {
        return things
            .groupBy(groupingFunction)
            .mapValues { (_, values) -> values.reduce(mergingFunction) }
            .toSortedMap(serializableNaturalOrder())
    }

    // G E N E R I C   M E R G I N G   F U N C T I O N S

    /**
     * Supply 'keep' merging function
     *
     * @param V type of element
     */
    private fun <V> keepMerging() = { existing: V, replacement: V ->
        if (existing == replacement) {
            if (LOG_DUPLICATE_VALUES) {
                Tracing.psInfo.println("[W] Duplicate values $existing and $replacement, keeping first")
            }
            //throw new IllegalArgumentException(existing + "," + replacement);
        }
        existing
    }

    /**
     * Supply 'replace' merging function
     *
     * @param V type of element
     */
    private fun <V> replaceMerging() = { existing: V, replacement: V ->
        if (existing == replacement) {
            if (LOG_DUPLICATE_VALUES) {
                Tracing.psInfo.println("[W] Duplicate values $existing and $replacement, replacing first")
            }
            //throw new IllegalArgumentException(existing + "," + replacement);
        }
        replacement
    }

    // S E N S E   A N D   S Y N S E T  M A P   F A C T O R I E S

    /**
     * Senses by id
     *
     * @receiver senses
     * @return senses mapped by id
     */
    fun Sequence<Sense>.sensesById(): Map<SenseKey, Sense> {
        // prioritize cased
        val mergingFunction = { existing: Sense, replacement: Sense ->
            val merged = if (replacement.lex.isCased) (if (existing.lex.isCased) existing else replacement) else existing
            if (existing == replacement) {
                if (LOG_DUPLICATE_VALUES) {
                    Tracing.psInfo.println("[W] Duplicate values $existing and $replacement, merging to $merged")
                }
            }
            merged
        }
        return map(this, { it.senseKey }, mergingFunction)
    }

    /**
     * Synsets by id
     *
     * @receiver synsets
     * @return synsets mapped by id
     */
    fun Sequence<Synset>.synsetsById(): Map<SynsetId, Synset> {
        return map(this) { it.synsetId }
    }
}
